### PR TITLE
 add insert_element function

### PR DESCRIPTION
**Pull request by kryv**
add insert_element function

This function returns new FLAME machine object.

I have add this function to "core/element.py".  Is there any other proper module file to add?

### CODE REVIEW

File: flame_utils/misc/confupdate.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import logging
+from flame import Machine
+
+import numpy as np
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def conf_update(machine):
```
Comment (archman) [inline flame_utils/misc/confupdate.py]: Why do you need this function?

Comment (kryv) [reply flame_utils/misc/confupdate.py]: *m.conf()* returns initial settings of the lattice.
If user change the lattice setting by using *reconfigure(i, setting)*,
*m.conf(i)* returns new setting, but *m.conf()* returns initial setting.
So, *conf_update()* apply current settings to the initial settings.

Comment (archman) [reply flame_utils/misc/confupdate.py]: I've been aware of that this function has some common part with `generate_latfile`, but not the same for handling duplicated elements?

Comment (kryv) [reply flame_utils/misc/confupdate.py]: You're right.
I think, I can polish the part of `generate_latfile` by using this function.

Comment (archman) [reply flame_utils/misc/confupdate.py]: Additonal note: update `ModelFlame` class with `insert_element` method and correct `clone_machine` method.

File: flame_utils/misc/confupdate.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    """Update machine conf() by using current settings.
+
+    :return: FLAME machine object
```
Comment (archman) [inline flame_utils/misc/confupdate.py]:
```
Returns
-------
```

File: flame_utils/core/model.py [hunk 3]
```diff
@@ -497,4 +499,29 @@ def clone_machine(self):
...
+    def insert_element(self, econf=None, index=None, element=None):
+        """Insert new element to the machine.
+
+        Parameters
+        ----------
+        econf : dict
+            Element configuration (see :func:`get_element`).
+        index : int or str
+            Insert element before the index (or element name).
+        element : dict
```
Comment (archman) [inline flame_utils/core/model.py]: Put an example of `element` dict, e.g. `{'name':xxx, 'length':xxxx}`.

File: flame_utils/core/element.py [hunk 4]
```diff
@@ -450,3 +451,37 @@ def get_names_by_pattern(pattern='.*', latfile=None, _machine=None):
...
+def insert_element(machine=None, index=None, element=None):
+    """Insert new element to the machine.
+
+    Parameters
+    ----------
+    machine :
+        FLAME machine object.
+    index :
+        Insert element before the index (or element name).
+    element :
```
Comment (archman) [inline flame_utils/core/element.py]: Put an example of `element` dict, e.g. `{'name':xxx, 'length':xxxx}`.

File: flame_utils/core/model.py [hunk 5]
```diff
@@ -497,4 +499,29 @@ def clone_machine(self):
         ret :
             FLAME Machine object.
         """
-        return Machine(self._mach_ins.conf())
+        return Machine(conf_update(self._mach_ins.conf()))
+
+    def insert_element(self, econf=None, index=None, element=None):
```
Comment (archman) [inline flame_utils/core/model.py]: I would put `econf` parameter to be the last one.

File: tests/test_core.py [hunk 6]
```diff
@@ -312,3 +312,19 @@ def test_generate_source(self):
...
+class TestInsertElemInModelFlame(unittest.TestCase):
+    def setUp(self):
+        testfile = os.path.join(curdir, 'lattice/test_0.lat')
+        self.testfile = make_latfile(testfile)
+
+    def test_insert_in_modelflame(self):
+        latfile = self.testfile
+        fm = ModelFlame(latfile)
+        new_econf = {'index':5, 'properties':{'name':'test_drift', 'type':'drift', 'L':1.0}}
+        fm.insert_element(new_econf)
```
Comment (archman) [inline tests/test_core.py]: If `econf` is not the first positional parameter, then `econf=` should be added.

File: tests/test_core.py [hunk 7]
```diff
@@ -312,3 +312,19 @@ def test_generate_source(self):
...
+        test_econf = fm.get_element(index=5)[0]
+        self.assertEqual(test_econf['index'], new_econf['index'])
```
Comment (archman) [inline tests/test_core.py]: Add a new test line for the element after the inserted one, whose index number should be changed from 5 to 6?

File: tests/test_core.py [hunk 8]
```diff
@@ -312,3 +312,19 @@ def test_generate_source(self):
...
+        self.assertEqual(test_econf['properties']['name'], new_econf['properties']['name'])
+        self.assertEqual(test_econf['properties']['type'], new_econf['properties']['type'])
+        self.assertEqual(test_econf['properties']['L'], new_econf['properties']['L'])
```
Comment (archman) [inline tests/test_core.py]: Maybe another test should be added to ensure that the new lattice can produce correct results?

File: tests/test_core.py [hunk 9]
```diff
@@ -312,3 +312,30 @@ def test_generate_source(self):
...
+class TestInsertElemInModelFlame(unittest.TestCase):
+    def setUp(self):
+        testfile = os.path.join(curdir, 'lattice/test_0.lat')
+        self.testfile = make_latfile(testfile)
+
+    def test_insert_in_modelflame(self):
+        latfile = self.testfile
+        fm = ModelFlame(latfile)
+        r0,s0=fm.run(to_element=6)
+
```
Comment (archman) [inline tests/test_core.py]: To this line, what is the result of `fm.get_element(index=5)`?

Comment (kryv) [reply tests/test_core.py]: The result of `fm.get_element(index=5)` in that point is,
`{name:'LS1_CA01:BPM_D1129', 'type': 'bpm'}`.
So, I checked `index=6` element parameters after the insertion is the same as the above.
https://github.com/kryv/flame-utils/blob/5f2e31a9065813cee240c6d013b884680fbf8baf/tests/test_core.py#L334-L337

Comment (archman) [reply tests/test_core.py]: I would suggest to save `fm.get_element(index=5)[0]` as a `econf_before_insertion`, then refactor line 336-337 by replacing hardcoded strings.

File: tests/test_core.py [hunk 10]
```diff
@@ -312,3 +312,30 @@ def test_generate_source(self):
...
+        self.assertEqual(test_econf['properties']['L'], new_econf['properties']['L'])
+
+        test_econf2 = fm.get_element(index=6)[0]
+        self.assertEqual(test_econf2['index'], 6)
+        self.assertEqual(test_econf2['properties']['name'], 'LS1_CA01:BPM_D1129')
+        self.assertEqual(test_econf2['properties']['type'], 'bpm')
+
+        r1,s1 = fm.run(to_element=6)
+
+        compare_mstates(self, s0, s1)
```
Comment (archman) [inline tests/test_core.py]: Is the inserted drift element the same as the original element with `index=6` before insertion? Or `s0` will not be equal to `s1`

Comment (kryv) [reply tests/test_core.py]: Yes. I have inserted the same length drift. At the first time, I have inserted the drift with arbitrary length and adjust the next drift length. But it causes the very small beam state difference due to the numerical error. In the `compare_mstates`, it request the all beam state value are completely the same. So I have adopted this routine.

**Comment by kryv** [comment]
In the case of insert after the last element, you can insert at lattice length. 
like,
```
>>> l = [1, 2]
>>> l.insert(2, 3)
>>> l
[1, 2, 3]
```

**Pull request merged**